pkg/ipam/ipam.proto

```diff
@@ -1,76 +1,62 @@
 syntax="proto3";
 option go_package = "./ipam";
 
-service ipam {
-    rpc MapEndpointIP (MapRequest) returns (MapResponse);
-    rpc UnmapEndpointIP (UnmapRequest) returns (UnmapResponse);
-    rpc MapNetworkCIDR (MapCIDRRequest) returns (MapCIDRResponse);
-    rpc UnmapNetworkCIDR (UnmapCIDRRequest) returns (UnmapCIDRResponse);
-    rpc GetHomePodIP (GetHomePodIPRequest) returns (GetHomePodIPResponse);
-    rpc BelongsToPodCIDR (BelongsRequest) returns (BelongsResponse);
-    rpc GetOrSetExternalCIDR(GetOrSetExtCIDRRequest) returns (GetOrSetExtCIDRResponse);
-    rpc SetSubnetsPerCluster(SetSubnetsPerClusterRequest) returns (SetSubnetsPerClusterResponse);
+service IPAM {
+    rpc IPAcquire (IPAcquireRequest) returns (IPAcquireResponse);
+    rpc IPRelease (IPReleaseRequest) returns (IPReleaseResponse);
+
+    rpc NetworkAcquire (NetworkAcquireRequest) returns (NetworkAcquireResponse);
+    rpc NetworkRelease (NetworkReleaseRequest) returns (NetworkReleaseResponse);
+    rpc NetworkIsAvailable (NetworkAvailableRequest) returns (NetworkAvailableResponse);
 }
 
-message MapRequest {
-    string clusterID = 1;
-    string ip = 2;
+message ResponseResult {
+    bool success = 1;
+    string error = 2;
 }
 
-message MapResponse {
+message IPAcquireRequest {
     string ip = 1;
+    string cidr = 2;
 }
 
-message UnmapRequest {
-    string clusterID = 1;
+message IPAcquireResponse {
+    ResponseResult result = 1;
     string ip = 2;
 }
 
-message UnmapResponse {}
-
-message MapCIDRRequest {
-    string cidr = 1;
+message IPReleaseRequest {
+    string ip = 1;
+    string cidr = 2;
 }
 
-message MapCIDRResponse {
-    string cidr = 1;
+message IPReleaseResponse {
+    ResponseResult result = 1;
 }
 
-message UnmapCIDRRequest {
+message NetworkAcquireRequest {
     string cidr = 1;
+    bool immutable = 2; // If true, the network cannot be remapped. It will be allocated if available, or an error will be returned.
 }
 
-message UnmapCIDRResponse {}
-
-message GetHomePodIPRequest {
-    string clusterID = 1;
-    string ip = 2;
-}
-
-message GetHomePodIPResponse {
-    string homeIP = 1;
-}
-
-message BelongsRequest {
-    string ip = 1;
+message NetworkAcquireResponse {
+    ResponseResult result = 1;
+    string cidr = 2;
 }
 
-message BelongsResponse {
-    bool belongs = 1;
+message NetworkReleaseRequest {
+    string cidr = 1;
 }
 
-message GetOrSetExtCIDRRequest {
-    string desiredExtCIDR = 1;
+message NetworkReleaseResponse {
+    ResponseResult result = 1;
 }
 
-message GetOrSetExtCIDRResponse {
-    string remappedExtCIDR = 1;
+message NetworkAvailableRequest {
+    string cidr = 1;
 }
 
-message SetSubnetsPerClusterRequest {
-    string remappedPodCIDR = 1;
-    string remappedExternalCIDR = 2;
-    string clusterID = 3;
+message NetworkAvailableResponse {
+    ResponseResult result = 1;
+    bool available = 2;
 }
```

Comment:
isn't the `available` boolean redundant if there is already one in the `ResponseResult`? I get that they can have different meanings, but in this case it seems redundant and not very clear

Reply:
This function checks if the network is available or not; in the resource result, we store the errors. We can have no errors and both available or not available networks, they are different things

Reply:
My question is more about the first boolean in the ResponseResult. What does it tell? If I am understanding correctly, I would just check if there is an error first, and then check the `available` bool, thus always discarding the first `success` bool

Reply:
The first success is like to check `err != nil`, while this one tells you if the network is available or not